cmd/ipfs/daemon.go

```diff
@@ -461,3 +482,80 @@ func merge(cs ...<-chan error) <-chan error {
 	}()
 	return out
 }
+
+func runGC(req cmds.Request, repo repo.Repo, node *core.IpfsNode) error {
+	disableGC, _, err := req.Option(disableGCKwd).Bool()
+	if err != nil {
+		return err
+	}
+	if disableGC {
+		return nil
+	}
+
+	cfg, err := req.InvocContext().GetConfig()
+	if err != nil {
+		return fmt.Errorf("runGC: GetConfig() failed: %s", err)
+	}
+	ctx := req.Context()
+
+	// check if cfg has these fields initialized
+	// TODO: there should be a general check for all of the cfg fields
+	// maybe distinguish between user config file and default struct?
+	if cfg.Datastore.StorageMax == 0 {
+		repo.SetConfigKey("Datastore.StorageMax", 10*1000)
+		cfg.Datastore.StorageMax = 10 * 1000
+	}
+	if cfg.Datastore.StorageGCWatermark == 0 {
+		repo.SetConfigKey("Datastore.StorageGCWatermark", 90)
+		cfg.Datastore.StorageGCWatermark = 90
+	}
+	if cfg.Datastore.GCTimeoutMinute == 0 {
+		repo.SetConfigKey("Datastore.GCTimeoutMinute", 60)
+		cfg.Datastore.GCTimeoutMinute = 60
+	}
+
+	// calculate "swap" size, the slack space between StorageMax and StorageGCWatermark
+	// used to limit GC duration
+	swapGB := cfg.Datastore.StorageMax / 1000 * (100 - cfg.Datastore.StorageGCWatermark) / 100
+	if swapGB < 1 {
+		swapGB = 1
+	}
```

Comment:
maybe let's rename this `slack`, as `swap` has other (incorrect) connotations (swap files, etc).